File: src/main/java/cu/suitetecsa/sdk/nauta/utils/DocumentUtils.java
package cu.suitetecsa.sdk.nauta.utils;

import cu.suitetecsa.sdk.nauta.exception.NautaException;
import cu.suitetecsa.sdk.nauta.exception.NotLoggedInException;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.util.List;

public class DocumentUtils {
    private static String getErrorMessage(Document html, PortalManager portalManager) {
        return new HtmlErrorParser().whenPortalManager(portalManager).parseError(html);
    }

    private static List<String> parseErrors(String errorMessage) {
        if (errorMessage.startsWith("Se han detectado algunos errores.")) {
            Document errorsHtml = Jsoup.parse(errorMessage);
            Elements subMessages = errorsHtml.select("li[class='sub-message']");
            return subMessages.stream().map(Element::text).toList();
        } else {
            return List.of(errorMessage);
        }
    }

    public static void throwExceptionOnFailure(Document html, String message, PortalManager portalManager) throws NotLoggedInException, NautaException {
        String errorMessage = getErrorMessage(html, portalManager);
        if (errorMessage != null) {
            List<String> errors = parseErrors(errorMessage);

            if (errors.size() == 1 && Constants.notLoggedInErrors.contains(errors.get(0))) {
                throw new ExceptionHandler.Builder<NotLoggedInException>()
                        .build()
                        .handleException(message, errors);
            }

            throw new ExceptionHandler.Builder<NautaException>().build().handleException(message, errors);
        }
    }

    public static <T extends Exception> void throwExceptionOnFailure(Document html, String message, PortalManager portalManager, ExceptionHandler<T> exceptionHandler) throws NautaException, T {
        String errorMessage = getErrorMessage(html, portalManager);
        if (errorMessage != null) {
            List<String> errors = parseErrors(errorMessage);

            if (errors.size() == 1 && Constants.notLoggedInErrors.contains(errors.get(0))) {
                throw new ExceptionHandler.Builder<NautaException>()
                        .build()
                        .handleException(message, errors);
            }

            throw exceptionHandler.handleException(message, errors);
        }
    }
}

File: src/main/kotlin/io/github/suitetecsa/sdk/nauta/utils/DocumentUtils.kt
package io.github.suitetecsa.sdk.nauta.utils

import io.github.suitetecsa.sdk.nauta.exception.NautaException
import io.github.suitetecsa.sdk.nauta.exception.NotLoggedInException
import io.github.suitetecsa.sdk.nauta.utils.HtmlErrorParser.parseError
import org.jsoup.Jsoup
import org.jsoup.nodes.Document

object DocumentUtils {
    private fun getErrorMessage(html: Document): String? {
        return parseError(html)
    }

    private fun parseErrors(errorMessage: String): List<String> {
        if (errorMessage.startsWith("Se han detectado algunos errores.")) {
            val errorsHtml = Jsoup.parse(errorMessage)
            val subMessages = errorsHtml.select("li[class='sub-message']")
            return subMessages.map { it.text() }
        } else {
            return listOf(errorMessage)
        }
    }

    /**
     * Throws a NotLoggedInException or NautaException if an error is detected in the HTML.
     *
     * @param html The HTML document to check for errors
     * @param message The error message to use if throwing an exception
     * @throws NotLoggedInException If a not logged in error is detected
     * @throws NautaException For any other errors detected
     */
    @JvmStatic
    @Throws(NotLoggedInException::class, NautaException::class)
    fun throwExceptionOnFailure(html: Document, message: String) = getErrorMessage(html)?.let {
            val errors = parseErrors(it)

            if (errors.size == 1 && Constants.notLoggedInErrors.contains(errors[0])) {
                throw ExceptionHandler.Builder<NotLoggedInException>()
                    .build()
                    .handleException(message, errors)
            }

            throw ExceptionHandler.Builder<NautaException>().build().handleException(
                message, errors
            )
        }

    @JvmStatic
    @Throws(NautaException::class)
    fun <T : Exception> throwExceptionOnFailure(
        html: Document,
        message: String?,
        exceptionHandler: ExceptionHandler<T>
    ) {
        val errorMessage = getErrorMessage(html)
        if (errorMessage != null) {
            val errors = parseErrors(errorMessage)

            if (errors.size == 1 && Constants.notLoggedInErrors.contains(errors[0])) {
                throw ExceptionHandler.Builder<NautaException>()
                    .build()
                    .handleException(message!!, errors)
            }

            throw exceptionHandler.handleException(message!!, errors)
        }
    }
}
